extensions/sql-bindings/src/test/service/azureFunctionsService.test.ts

```diff
@@ -0,0 +1,201 @@
+		sinon.stub(utils, 'getVscodeMssqlApi').resolves(testUtils.vscodeMssqlIExtension.object);
+
+		let connectionInfo: IConnectionInfo = createTestCredentials();// Mocks promptForConnection
+		connectionInfo.database = 'testDb';
+
+		let connectionDetails = { options: connectionInfo };
+		testUtils.vscodeMssqlIExtension.setup(x => x.getConnectionString(connectionDetails, true, false)).returns(() => Promise.resolve('testConnectionString'));
+
+		const spy = sinon.spy(vscode.window, 'showErrorMessage');
+
+		// select input or output binding
+		let quickpickStub = sinon.stub(vscode.window, 'showQuickPick').resolves(<any>{ label: constants.input, type: BindingType.input });
+
+		// no table used for connection info so prompt user to get connection info
+		testUtils.vscodeMssqlIExtension.setup(x => x.promptForConnection(true)).returns(() => Promise.resolve(connectionInfo));
+		testUtils.vscodeMssqlIExtension.setup(x => x.connect(connectionInfo)).returns(() => Promise.resolve('testConnectionURI'));
+		testUtils.vscodeMssqlIExtension.setup(x => x.listDatabases('testConnectionURI')).returns(() => Promise.resolve(['testDb']));
+		// select the testDB from list of databases based on connection info
+		quickpickStub.onSecondCall().returns(Promise.resolve('testDb') as any);
+		// get tables from selected database
+		const params = { ownerUri: 'testConnectionURI', queryString: azureFunctionUtils.tablesQuery('testDb') };
+		testUtils.vscodeMssqlIExtension.setup(x => x.sendRequest(azureFunctionsContracts.SimpleExecuteRequest.type, params))
+			.returns(() => Promise.resolve({ rowCount: 1, columnInfo: [], rows: [['[schema].[testTable]']] }));
+		// select the schema.testTable from list of tables based on connection info and database
+		quickpickStub.onThirdCall().returns(Promise.resolve('[schema].[testTable]') as any);
+
+		// set azure function name
+		let inputStub = sinon.stub(vscode.window, 'showInputBox').resolves('testFunctionName');
+
+		// promptAndUpdateConnectionStringSetting
+		quickpickStub.onCall(3).resolves(<any>{ label: constants.createNewLocalAppSettingWithIcon });
+		inputStub.onSecondCall().resolves('SqlConnectionString');
+		// promptConnectionStringPasswordAndUpdateConnectionString - tested in AzureFunctionUtils.test.ts
+		quickpickStub.onCall(4).returns(Promise.resolve(constants.yesString) as any);
+		testUtils.vscodeMssqlIExtension.setup(x => x.getConnectionString(connectionDetails, true, false)).returns(() => Promise.resolve('testConnectionString'));
+		// setLocalAppSetting with connection string setting name and connection string
+		// fails if we dont set writeFile stub
+		sinon.stub(fs.promises, 'writeFile');
+		sinon.stub(azureFunctionUtils, 'setLocalAppSetting').withArgs(sinon.match.any, 'SqlConnectionString', 'testConnectionString').returns(Promise.resolve(true));
+		sinon.stub(utils, 'executeCommand').resolves('downloaded nuget package');
+
+		let testWatcher = vscode.workspace.createFileSystemWatcher((
```

Comment:
Do you actually need to create a watcher? It would be best to use typemoq to make a mocked version of it if possible so we don't end up with potential weird issues. In general unit tests should always try to avoid interacting with external systems as much as possible (such as the file system)

(same for the other watchers)